### PR TITLE
[inductor] Decompose small dot-shaped bmm

### LINKED ISSUE #176706: torch.dot under vmap lowers to unfused extern_kernels.bmm — pointwise mul+sum is 25-30% faster for small vectors
## 🐛 Describe the bug

When `torch.dot(a, b)` is used inside `torch.vmap` and compiled with `torch.compile`, it gets batched into `torch.bmm` and lowered to `extern_kernels.bmm` (cuBLAS). For small vectors (3-4 elements, typical in physics simulations), this is significantly slower than the equivalent `(a * b).sum(-1)` because:

1. **cuBLAS dispatch overhead** dominates for tiny matrices like `(B, 1, 3) @ (B, 3, 1)`
2. **Fusion barrier** — `extern_kernels.bmm` is opaque to Inductor, so surrounding pointwise ops cannot be fused into the same Triton kernel
3. **Kernel count explosion** — each dot product becomes a standalone kernel launch

## Benchmark evidence

In [mujoco-torch](https://github.com/vmoens/mujoco-torch) (a PyTorch physics simulator), replacing every `torch.dot` call with `(a * b).sum(-1)` yielded **+25–30% throughput improvement** in `torch.compile` mode (H200, float64, B=32768). The codebase had ~40 `torch.dot` call sites on 3-element and 4-element vectors (positions, quaternion components).

Before the change, the compiled Inductor output for a single model contained **83 `extern_kernels.bmm` calls**, most operating on shapes like `(32768, 1, 3) @ (32768, 3, 1)`. After the change, these were all fused into neighboring Triton kernels.

## Minimal repro

```python
import torch

def dot_based(a, b):
    # 10 chained dot products (simulating a physics pipeline)
    s = torch.dot(a, b)
    for _ in range(9):
        s = s + torch.dot(a, b)
    return s

def pointwise_based(a, b):
    s = (a * b).sum(-1)
    for _ in range(9):
        s = s + (a * b).sum(-1)
    return s

B = 32768
a = torch.randn(B, 3, dtype=torch.float64, device="cuda")
b = torch.randn(B, 3, dtype=torch.float64, device="cuda")

vmap_dot = torch.vmap(dot_based)
vmap_pw  = torch.vmap(pointwise_based)

compiled_dot = torch.compile(vmap_dot, fullgraph=True)
compiled_pw  = torch.compile(vmap_pw, fullgraph=True)

# Warmup
for _ in range(3):
    compiled_dot(a, b)
    compiled_pw(a, b)
torch.cuda.synchronize()

import time

torch.cuda.synchronize()
t0 = time.perf_counter()
for _ in range(1000):
    compiled_dot(a, b)
torch.cuda.synchronize()
t_dot = time.perf_counter() - t0

torch.cuda.synchronize()
t0 = time.perf_counter()
for _ in range(1000):
    compiled_pw(a, b)
torch.cuda.synchronize()
t_pw = time.perf_counter() - t0

print(f"torch.dot:      {t_dot:.3f}s")
print(f"mul+sum:         {t_pw:.3f}s")
print(f"speedup:         {t_dot / t_pw:.2f}x")
```

## Inductor output comparison

**With `torch.dot`** (via vmap → bmm):
```python
extern_kernels.bmm(reinterpret_tensor(buf42, (32768, 1, 3), ...), 
                    reinterpret_tensor(buf43, (32768, 3, 1), ...), 
                    out=buf44)
```
Each call is a standalone cuBLAS invocation that breaks the fusion chain.

**With `(a * b).sum(-1)`:**
The multiply and sum are pointwise/reduction ops that Inductor fuses into the surrounding Triton kernel — no separate kernel launch.

## Suggested fix

Several possible approaches (not mutually exclusive):

1. **`vmap` batching rule for `dot`**: when the contraction dimension is small (e.g. ≤ 8), emit `(a * b).sum(-1)` instead of `bmm`
2. **Inductor lowering**: recognize tiny `bmm` (inner dim ≤ threshold) and lower to pointwise multiply + reduce instead of cuBLAS
3. **Inductor fusion pass**: allow small `extern_kernels.bmm` to be inlined as fused Triton ops

Option 2 seems most general — it would benefit any code path that produces small bmm, not just vmap'd dot products.

## Versions

- PyTorch: built from source (2.12.0a0+git8a4793d)
- GPU: NVIDIA H200
- CUDA: 12.x
- OS: Linux

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #183911

Lower CUDA/XPU bmm with shape (B, 1, K) @ (B, K, 1) and small K to a fused multiply/reduction when the dtype matches eager bmm support, avoiding tiny extern bmm launches from vmap(dot).

Fixes #176706

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

File: torch/_inductor/kernel/bmm.py [hunk 1]
```diff
@@ -75,12 +75,41 @@ def bmm_grid(b, m, n, meta, *, cdiv, max):
...
+        and device_type == mat2.get_device().type
+        and dtype == mat2.get_dtype()
+        and dtype in _BMM_DOT_DECOMPOSE_DTYPES
+        and sizevars.statically_known_equals(mat1.get_size()[1], 1)
+        and sizevars.statically_known_equals(mat2.get_size()[2], 1)
+        and sizevars.statically_known_leq(
+            mat1.get_size()[2], _BMM_DOT_K_DECOMPOSE_THRESHOLD
+        )
+    ):
+        mat1 = L.unsqueeze(mat1, -1)
```
Comment (aorenste) [inline torch/_inductor/kernel/bmm.py:122]: nit: Could use a trivial comment to explain why we're doing this.

File: torch/_inductor/kernel/bmm.py [hunk 2]
```diff
@@ -75,12 +75,41 @@ def bmm_grid(b, m, n, meta, *, cdiv, max):
     torch.baddbmm, "at::baddbmm_out", op_overload=aten.baddbmm.out
 )
 
+_BMM_DOT_K_DECOMPOSE_THRESHOLD = 32
```
Comment (aorenste) [inline torch/_inductor/kernel/bmm.py:82]: Is this value a guess? Empirically tested? Is it the kind of thing we'll want to update over time (in which case how do we test)?

File: test/inductor/test_torchinductor.py [hunk 3]
```diff
@@ -4348,6 +4348,72 @@ def fn(a, b):
...
+    @skip_if_not_triton
+    def test_vmap_dot_decomposes_bmm(self):
+        def dot_based(a, b):
+            return torch.dot(a, b) + torch.dot(a, b)
+
+        fn = torch.vmap(dot_based)
+        bmm_codegen_call = (
+            "aoti_torch_cuda_bmm_out" if config.cpp_wrapper else "extern_kernels.bmm"
+        )
+        for dtype in (torch.float32, torch.float64):
```
Comment (aorenste) [inline test/inductor/test_torchinductor.py]: This doesn't cover all the types in _BMM_DOT_DECOMPOSE_DTYPES, should it?

File: torch/_inductor/kernel/bmm.py [hunk 4]
```diff
@@ -75,12 +75,41 @@ def bmm_grid(b, m, n, meta, *, cdiv, max):
...
+    sizevars = V.graph.sizevars
+    dtype = mat1.get_dtype()
+    device_type = mat1.get_device().type
+    if (
+        out_dtype is None
+        and device_type in ("cuda", "xpu")
+        and device_type == mat2.get_device().type
+        and dtype == mat2.get_dtype()
+        and dtype in _BMM_DOT_DECOMPOSE_DTYPES
+        and sizevars.statically_known_equals(mat1.get_size()[1], 1)
```
Comment (aorenste) [inline torch/_inductor/kernel/bmm.py]: Drive by: Should the CPU check below (not this PR) also check statically_known_equals like this rather than (I guess) adding a guard? (cc @laithsakka)

**Review by aorenste — APPROVED** [review]
Approving w/ comments

**Comment by jansel** [comment]
@pytorchbot merge

(I checked that the agent added requested comments and updated tests)

**Comment by izaitsevfb** [comment]
@pytorchbot revert -m "looks like the new tests aren't compatible with dynamic shapes" -c nosignal



test/inductor/test_torchinductor_codegen_dynamic_shapes.py::DynamicShapesCodegenGPUTests::test_bmm_dot_shape_decompose_threshold_dynamic_shapes_cuda [GH job link](https://github.com/pytorch/pytorch/actions/runs/26117445705/job/76825518179) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/4509de7b1517425c57640cfa6ffded7e4381511d)
test/inductor/test_torchinductor_dynamic_shapes.py::DynamicShapesGPUTests::test_vmap_dot_decomposes_bmm_dynamic_shapes_cuda [GH job link](https://github.com/pytorch/pytorch/actions/runs/26117445705/job/76825518170) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/4509de7b1517425c57640cfa6ffded7e4381511d)

test/inductor/test_torchinductor_dynamic_shapes.py::DynamicShapesGPUTests::test_bmm_dot_shape_decompose_threshold_dynamic_shapes_cuda [GH job link](https://github.com/pytorch/pytorch/actions/runs/26117445705/job/76825518300) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/4509de7b1517425c57640cfa6ffded7e4381511d)

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes

File: torch/_inductor/kernel/bmm.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/kernel/bmm.py:122]: My agent says this is addressed in the current diff with comments before the threshold and before the dot-shaped bmm decomposition.

File: torch/_inductor/kernel/bmm.py (hunk 2 above)
Comment (jansel) [reply torch/_inductor/kernel/bmm.py:82]: My agent says the threshold is intentionally conservative for tiny vector contractions; the current tests cover both the K=32 and K=33 boundary so future changes have a regression check.

**Comment by jansel** [comment]
@claude review these changes

File: torch/_inductor/kernel/bmm.py (hunk 4 above)
Comment (jansel) [reply torch/_inductor/kernel/bmm.py]: My agent says acknowledged; this was a drive-by note about the existing CPU path rather than this PR, so I kept the change scoped to the CUDA/XPU small dot-shaped bmm path and left the CPU behavior unchanged.

**Comment by jansel** [comment]
Agent, confirm that the reason for the revert has been resolved.  Is this safe to reland?  Also confirm that the problem being fixed still exists without this change.

**Comment by jansel** [comment]
My agent says yes: the revert was for dynamic-shape bmm/vmap-dot tests, and I reran those exact dynamic-shape tests plus the codegen variant on the PR tree; all pass. I also verified current main without this PR still emits extern_kernels.bmm for torch.compile(torch.vmap(torch.dot)) on CUDA (extern_kernels.bmm_count=1), so the original issue remains live without the change. Dr. CI says this can merge normally with only unrelated failures, so this is safe to reland.
